Added Pixel Hashing to Image Consolidation

BolWaffy opened the pull request:
The current `Combine Images` operator can accidentally merge different image textures, which can sometimes cause a headache. So I upgraded the operator to use MD5 Pixel Hashing and compare the actual image data to verify they are identical before merging.

Also added 'UNDO' to the bl_options so Blender does not crash if the user undoes that step.

BolWaffy committed:
Removed pre 2.78 logic

Removed outdated material assignment logic for Blender versions prior to 2.78.

BolWaffy committed:
Fix typos in material_manager.py descriptions

BolWaffy committed:
Added pixel hashing

TheDuckCow commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -285,90 +289,105 @@ class MCPREP_OT_combine_images(bpy.types.Operator):
 		removeold = True
 
-		if self.selection_only is True and len(context.selected_objects) == 0:
+		if bpy.app.version < (2, 78):
```
We already support only Blender 2.8+ so we can just remove this block (appreciate trying to maintain the functionality though)

TheDuckCow commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -285,90 +289,105 @@ class MCPREP_OT_combine_images(bpy.types.Operator):
+			# Reshape so buffer skips WHOLE pixels, keeping RGBA grouped
+			sampled = pixels.view().reshape(-1, 4)[::buffer]
+			
+			return hashlib.md5(sampled.tobytes()).hexdigest()
```
I had a high level over the speed concerns, so I did some timing tests. My results:

## Test script

```python
import time
t0 = time.time()
bpy.ops.mcprep.combine_images(skipUsage=True)
t1 = time.time()
print("Time in seconds: ", t1-t0)
```

## Testing the control, ie current dev branch

-- MC scene
Time in seconds:  0.02409982681274414
Time in seconds:  0.026123046875

Outcome: basically instant


-- General PBR scene (mixture of 1K and 2K assets, 95 images total across only ~13 materials)
Time in seconds:  0.013833999633789062
Time in seconds:  0.012299776077270508
Time in seconds:  0.012421846389770508

Outcome: basically instant

-- MC PBR scene (same as the MC one above, but swapped for a PBR pack; 224 initial images, )
Time in seconds:  0.1075751781463623
Time in seconds:  0.05999588966369629
Time in seconds:  0.06019115447998047

Outcome: basically instant


## Now with the new changes of this PR:


-- MC version:
Time in seconds:  0.03257012367248535
Time in seconds:  0.04587864875793457
Time in seconds:  0.0442049503326416

Outcome: still basically instant

-- General PBR scene (mixture of 1K and 2K assets, 95 images total across only ~13 materials)
Time in seconds:  3.233705997467041
Time in seconds:  3.362300157546997
Time in seconds:  3.2018487453460693

Outcome: Noticeably slower; didn't have any 8K assets so this could definitely end up getting even slower

-- MC PBR scene (same as the MC one above, but swapped for a PBR pack)
Time in seconds:  0.1335008144378662
Time in seconds:  0.12597393989562988
Time in seconds:  0.12844491004943848

Outcome: not instant, but fast enough


## Summary:

For typical MCprep user use cases, this is still fast enough, especially since it's not an operation that users are running that often. It can be quite slow if this were ever used in a larger PBR scene that uses higher resolution assets, but could be still mitigated by users working based off of selection.

TheDuckCow commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -285,90 +289,105 @@ class MCPREP_OT_combine_images(bpy.types.Operator):
+			
+		except Exception as e:
+			print(f"Failed to hash {img.name}: {e}")
+			return None
```
Code comment: let's avoid bare exceptions. If there can be an exception, let's try and focus the handling around the most narrow line(s) of code possible. Ideally, that would mean only 1 line of in a try/except block.

TheDuckCow reviewed (approved):
Approving, but making a small suggestion for the exception handling. Per my analysis below, the speed is indeed a little slower, but for most use cases, it will not be perceptible or will be largely quite tolerable. (I contemplated suggesting a secondary option in case users want to do a "fast" check, but this is not an operator that benefits from having multiple options in practice, and just increases complexity while being harder to understand).

StandingPadAnimations commented:
Now that I think about it, I think we could actually just compare file contents. After all, the file contents have to be read anyway to perform the hash operation, so we can simply compare numpy arrays and skip the hashing process. That would probably help with performance as well. 

(Also avoids the possibility of hash collisions, which although rare can still happen)

BolWaffy committed:
micro optimizations

BolWaffy commented:
Forgot it was New Year's yesterday. Anyway, I thought I found a way around not needing to load the image into memory, but I was incorrect. I just did a few micro optimizations. In my test, I think the code is 1.15x faster than the first edit.

BolWaffy committed:
Refactor get_image_hash to return byte-string hash

Updated the get_image_hash method to return a byte-string hash instead of an MD5 hash.

BolWaffy committed:
Remove Blender version requirement check

BolWaffy committed:
Adjusted the try/except block, little format cleanup

BolWaffy committed:
Added back group RGBA channels

This be the last "oh, I forgot that" edit.

BolWaffy commented:
I already have code that retrieves image data blocks from materials, so if you want, I can add the Combine by Selection option, which should be straightforward.

I can also implement a similar hashing comparison approach with the Combine Materials tool. I also already have code that converts the material node tree into JSON. The added execution time would be minimal (just a couple of extra milliseconds).

BolWaffy committed:
Enhance image consolidation functionality

Added options for grouping images by name, selection, and strict comparison for pixel data. Improved stability.

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -268,106 +270,150 @@ def getMaterials(self, context):
-		default=False,
-		options={'HIDDEN'})
+
+	strict_comparison: bpy.props.BoolProperty(
+		name="Strict Comparison",
+		description="Check every single pixel. Slower, but 100% accurate",
+		default=False)
```
I would rephrase the description as "Compare image contents, which while slower, will be more accurate"

Also @TheDuckCow given this is intended to solve an issue that exists, do you think this should be set to `True` by default?

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -268,106 +270,150 @@ def getMaterials(self, context):
+					if mat.use_nodes and mat.node_tree:
+						for node in mat.node_tree.nodes:
+							if node.type == 'TEX_IMAGE' and node.image:
+								target_images.add(node.image)
```
Stylistic comment: I would invert the condition to avoid extreme nesting.


```suggestion
							if node.type != 'TEX_IMAGE' or not node.image:
								continue
							target_images.add(node.image)
```

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -268,106 +270,150 @@ def getMaterials(self, context):
+		# Group images
+		groups = {}
+		for img in images_to_check:
+			if img.type in {'RENDER_RESULT', 'COMPOSITING'} or img.use_fake_user:
```
Since the comparison is against a static list of values, it's better to make this a tuple.


```suggestion
			if img.type in ('RENDER_RESULT', 'COMPOSITING') or img.use_fake_user:
```

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -268,106 +270,150 @@ def getMaterials(self, context):
+
+		images_to_remove = []
+
+		# Hashing and Remapping
```
Since hashing requires reading the image contents into the hash function anyway, it would be a lot simpler to just compare image contents. Also probably a lot faster, especially when using Numpy.

Hashing in general is really only advised if you need an identifier based on data, like if you have to make files. However, since we already have the image data, it's both simpler and (most likely) faster to check the data directly, especially if we use vectorized Numpy operations. It also completely eliminates the rare (but non-zero) possibility of hash collisions.

BolWaffy replied:
I’ve been attempting to implement the direct data comparison using Numpy as you suggested, but so far every Numpy based comparison approach I've tried has actually been slower. I may be missing something obvious with how to structure the comparison efficiently.

StandingPadAnimations replied:
Looking at the code, I think the main issue is dictionaries that use the image contents as the key, and using keys for comparison. I've done some refactoring that's closer to what I meant by direct comparison (attached patch file below, can be applied with `git-apply`)

[comparision-refactor.patch](https://github.com/user-attachments/files/24501782/comparision-refactor.patch)

StandingPadAnimations reviewed (changes requested):
Mostly stylistic comments, but one fundamental change is needed: instead of hashing, it would be simpler (and most likely faster) to compare image contents directly with Numpy. This would avoid the calculation time for a hash on image data that we already have, and it would avoid the possibility of hash collisions.

BolWaffy committed:
Small fixes

BolWaffy committed:
Changes to descriptions

BolWaffy committed:
Micro optimization additions

BolWaffy committed:
Remove timing print statement from execute method

Removed timing print statement from execute method.

BolWaffy committed:
Refactor image handling and remove get_image_hash method

StandingPadAnimations committed:
refactor: Reduce use of dictionaries in image consolidation

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

BolWaffy committed:
Refactor image comparison, added quick prefilter and small adjustment's

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -331,64 +331,69 @@ def execute(self, context):
+
+			num_pixels = w * h * 4
+			img_array = np.empty(num_pixels, dtype=np.float32)
+			img.pixels.foreach_get(img_array)	# faster than `np.asarray(img.pixels)`
```
Image pixels in Blender convert to `NDArray[np.float64]`. IMO to avoid precision loss, that datatype should be used.

Also do we have actual metrics for `img.pixels.foreach_get(img_array)` being faster than using `np.asarray(img.pixels)`? If not, the latter should be used to reduce room for error.

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -331,64 +331,69 @@ def execute(self, context):
+			for uni_image, uni_data, uni_sum in unique_images[cur_img_key]:
+				# FAST PREFILTER
+				if not pix_sum == uni_sum:
+					continue
```
Since we're dealing with (usually) small images, the sum of all pixels isn't a good metric to determine if 2 images are unique (especially since we cap it in the end anyway). Since strict comparison is supposed to actually check pixel contents, it would be preferable to do just that. Performance is important, but since this is currently opt-in, we don't need something as fast as the non-strict check.

If this is intended to run when strict check is disabled, then I would add another condition for checking if strict check is disabled.

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -331,64 +331,69 @@ def execute(self, context):
 		# Group images
-		#
-		# Here we use a list, since we're just trying to get the
-		# raw images at this point for comparison
```
Since this is an explanatory comment, I would avoid removing it

StandingPadAnimations commented on MCprep_addon/materials/material_manager.py:
```diff
@@ -331,64 +331,69 @@ def execute(self, context):
+			stride = max(1, num_pixels // 16384) if not self.strict_comparison else 1
+			cur_img_data = img_array[::stride] if stride > 1 else img_array
+
+			pix_sum = np.sum(cur_img_data)
```
Since this does seem to be intended to run only when strict check is disabled, I would wrap this around a condition:

```py
pix_sum = None
if self.strict_comparision:
	stride = max(1, num_pixels)
	cur_img_data = img_array[::stride] if stride > 1 else img_array
	
	pix_sum = np.sum(cur_img_data)

if cur_img_key not in unique_images:
	unique_images[cur_img_key] = [(img, cur_img_data, pix_sum)]
```

In this case, I would also wrap the `float` type in `Optional`, and update later sections to check if the sum is `None`

```py
unique_images: Dict[Tuple[str, int, int], List[Tuple[bpy.types.Image, NDArray[np.float32], Optional[float]]]] = {}
```

StandingPadAnimations reviewed (changes requested):
A couple of small changes, which I'll make on my end and push. The main thing is making it clearer in code that the fast prefilter is only supposed to run when strict comparison is disabled.

I did make a note regarding array creation performance; as far as I'm aware, there shouldn't be any major performance issues with `np.asarray(img.pixels)`, so unless there's a major performance issue with it, I don't think there should be any practical issues.

StandingPadAnimations committed:
refactor: Put fast prefilter behind strict comparison checks

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations commented:
I'll add some tests for strict comparison (mostly just a copy-paste of the existing combine images test to make sure the code works), and then merge it

StandingPadAnimations committed:
tests: Add test for Combine Images with strict comparison

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations reviewed (approved):
Since tests past, and the code looks good, I'll merge this in
